In [1]:
import re
from pathlib import Path
import random
from collections import Counter
import math

### Data Loading and Pre-processing

This section handles loading the TREC06p dataset (a spam/ham email classification dataset) from the specified root directory. We define helper functions to create the dataset by reading labels and email contents, then preprocess the data for training. The create_dataset function reads email files with error handling for encoding issues, ensuring robustness. This step is crucial for preparing raw text data into lists of features (X) and labels (y), where X contains full email bodies and y contains binary labels ("spam" or "ham").

In [2]:
data_root = Path("../inputs/trec06p-ai201")

In [3]:
def dataset_split(X, y, test_size=0.3, random_state=42):
    random.seed(random_state)

    combined = list(zip(X, y))
    random.shuffle(combined)
    split = int(len(combined) * (1 - test_size))
    X_train, y_train = zip(*combined[:split])
    X_test, y_test = zip(*combined[split:])
    
    return list(X_train), list(X_test), list(y_train), list(y_test)

def create_dataset(data_root):
    with open(f"{data_root}/labels", "r") as f:
        labels, paths = zip(*(line.strip().split(maxsplit=1) for line in f))

    y = list(labels)
    paths = list(paths)

    X = []

    for path in paths:
        file = data_root.joinpath(path.replace("../", ""))
        with open(file, "r", encoding="utf-8", errors="ignore") as f:
            content = f.read()
            X.append(content)

    return X, y

### Train-Test split 

Here, we perform a random train-test split on the dataset using a custom dataset_split function. This ensures the model is trained on 70% of the data and evaluated on the remaining 30%, preventing overfitting. Random shuffling maintains balance in class distribution. The split is stored in X_train, X_test, y_train, and y_test for use in subsequent steps.

In [4]:
X,y = create_dataset(data_root)

X_train, X_test, y_train, y_test = dataset_split(X, y, test_size=0.3, random_state=42)

### Build Vocab using Training Data

We build a vocabulary set from unique words in the training emails using a custom word-level tokenizer. The tokenizer uses regex to extract alphabetic sequences, converting text to lowercase for consistency. We also compute word counts per class ("spam" and "ham") using a presence-based model (unique words per document), which helps in calculating likelihoods. The vocabulary size (V) is printed later, and this step lays the foundation for the Naive Bayes model by identifying relevant features.

In [5]:
def word_level_tokenize(text):
    # Word = alphabetic sequence, followed by space/comma/period
    pattern = r'\b[a-zA-Z]+(?=[\s,\.])'
    tokens = re.findall(pattern, text.lower())
    return [w for w in tokens]

In [6]:
vocab = set()
word_counts = {
    "spam": Counter(),
    "ham": Counter()
}

doc_counts = Counter(y_train)

for text, label in zip(X_train, y_train):
    words = set(word_level_tokenize(text))      
    vocab.update(words)
    word_counts[label].update(words)

V = sorted(vocab)


### Document counts per class

This section counts the number of training documents per class (D_spam and D_ham). These counts are essential for computing priors and denominators in the likelihood calculations. The output shows the class distribution, which can indicate dataset balance (e.g., if one class dominates, it may affect model bias).

In [7]:
D_spam = doc_counts["spam"]
D_ham = doc_counts["ham"]

print("D_spam:", D_spam)
print("D_ham:", D_ham)

D_spam: 17353
D_ham: 9122


### Compute priors

Priors represent the baseline probability of each class in the training data. We calculate P(spam) and P(ham) as the proportion of documents in each class. These are logged for numerical stability in the prediction phase. Priors help initialize scores in the Naive Bayes classifier, weighting predictions toward the more common class.



In [8]:
total_docs = D_spam + D_ham

prior = {
    "spam": D_spam / total_docs,
    "ham": D_ham / total_docs
}
print("Prior P(spam):", prior["spam"])
print("Prior P(ham):", prior["ham"])

Prior P(spam): 0.655448536355052
Prior P(ham): 0.34455146364494804


### Likelihood Calculation with Laplace Smoothing

The formulas show conditional probabilities P(x_i|spam) and P(x_i|ham) using Laplace smoothing (λ) to handle zero counts.

In [9]:
def train_naive_bayes(word_counts, V, D_spam, D_ham, lmbda=1):    
    log_likelihood = {
        "spam": {},
        "ham": {}
    }

    spam_denominator = D_spam + lmbda * len(V)
    ham_denominator = D_ham + lmbda * len(V)

    # Compute likelihoods with Laplace smoothing
    for word in V:
        count_spam = word_counts["spam"][word]
        count_ham = word_counts["ham"][word]

        log_likelihood_spam = math.log((count_spam + lmbda) / spam_denominator)
        log_likelihood_ham = math.log((count_ham + lmbda) / ham_denominator)

        log_likelihood["spam"][word] = log_likelihood_spam
        log_likelihood["ham"][word] = log_likelihood_ham

    log_prior = {
        "spam": math.log(D_spam / total_docs),
        "ham": math.log(D_ham / total_docs)
        }   

    model = {
        "log_likelihood": log_likelihood,
        "log_prior": log_prior,
        "lambda": lmbda,
        "vocab": V
    }

    return model

In [10]:
model = train_naive_bayes(word_counts, V, D_spam, D_ham, lmbda=1)

### Predict test emails

Prediction uses the log-probability scores for words in the email. The class with the higher score is chosen. We use a presence model (unique words) and ignore out-of-vocabulary words to avoid errors. This approach is computationally efficient for text classification.

In [11]:
def predict(text, model):
    words = set(word_level_tokenize(text))   # UNIQUE words (presence model)

    spam_score = model["log_prior"]["spam"]
    ham_score  = model["log_prior"]["ham"]

    ll_spam = model["log_likelihood"]["spam"]
    ll_ham  = model["log_likelihood"]["ham"]

    for w in words:
        if w in ll_spam:   # w is in training vocab
            spam_score += ll_spam[w]
            ham_score  += ll_ham[w]

    return "spam" if spam_score > ham_score else "ham"

In [12]:
def evaluate(X_test, y_test,
             model):
    TP = FP = FN = TN = 0

    for text, true_label in zip(X_test, y_test):
        pred = predict(text, model)

        if pred == "spam" and true_label == "spam":
            TP += 1
        elif pred == "spam" and true_label == "ham":
            FP += 1
        elif pred == "ham" and true_label == "spam":
            FN += 1
        elif pred == "ham" and true_label == "ham":
            TN += 1

    precision = TP / (TP + FP) if (TP + FP) > 0 else 0.0
    recall    = TP / (TP + FN) if (TP + FN) > 0 else 0.0
    f1_score  = 2 * (precision * recall) / (precision + recall) if (precision + recall) > 0 else 0.0

    return {
        "TP": TP, "FP": FP, "FN": FN, "TN": TN,
        "precision": precision,
        "recall": recall,
        "f1_score": f1_score
    }

### Evaluate precision / recall

Evaluation computes confusion matrix metrics (TP, FP, FN, TN) and derived scores: precision (TP/(TP+FP)), recall (TP/(TP+FN)), and F1-score (harmonic mean). These measure the model's ability to detect spam while minimizing false positives. Results are printed for the baseline model (λ=1).

In [13]:
results = evaluate(
    X_test, y_test,
    model
)

print(results)
print("Precision:", results["precision"])
print("Recall:", results["recall"])
print("f1_score:", results["f1_score"])

{'TP': 7512, 'FP': 72, 'FN': 47, 'TN': 3716, 'precision': 0.990506329113924, 'recall': 0.9937822463288795, 'f1_score': 0.9921415835699663}
Precision: 0.990506329113924
Recall: 0.9937822463288795
f1_score: 0.9921415835699663


### Train with different lambdas

We experiment with multiple λ values (2.0, 1.0, 0.5, 0.1, 0.005) to tune smoothing. Lower λ reduces smoothing, potentially improving fit but risking overfitting. Results show a precision-recall trade-off; λ=0.5 balances both, yielding the best F1-score.

In [14]:
lambdas = [2.0, 1.0, 0.5, 0.1, 0.005]

models = {}

for lmbda in lambdas:
    models[lmbda] = train_naive_bayes(word_counts, V, D_spam, D_ham, lmbda)


In [15]:
results = []

for lmbda in lambdas:
    model = models[lmbda]
    metrics = evaluate(X_test, y_test, model)

    print(f"Lambda: {lmbda}")
    print(f"Precision: {metrics['precision']}")
    print(f"Recall: {metrics['recall']}")
    print(f"F1 Score: {metrics['f1_score']}")
    print()

    results.append({
        "lambda": lmbda,
        "precision": metrics["precision"],
        "recall": metrics["recall"],
        "f1_score": metrics["f1_score"]
    })

Lambda: 2.0
Precision: 0.9883063986335567
Recall: 0.9951051726418838
F1 Score: 0.9916941331575477

Lambda: 1.0
Precision: 0.990506329113924
Recall: 0.9937822463288795
F1 Score: 0.9921415835699663

Lambda: 0.5
Precision: 0.9933774834437086
Recall: 0.9921947347532742
F1 Score: 0.9927857568336752

Lambda: 0.1
Precision: 0.9986515641855448
Recall: 0.9797592274110332
F1 Score: 0.9891151919866444

Lambda: 0.005
Precision: 0.9997225305216426
Recall: 0.9533007011509459
F1 Score: 0.9759599106114986



We evaluated five values of the smoothing parameter **λ: 2.0, 1.0, 0.5, 0.1, and 0.005**.
As λ decreases, precision increases while recall decreases, indicating a trade-off between avoiding false positives and detecting spam messages.

The value λ = 0.5 provides the best balance between **precision (0.9934) and recall (0.9922)**, achieving high performance on both metrics.

As additional note, at λ = 0.5 yields the best F1-Score.

**Therefore, λ = 0.5 is selected as the optimal smoothing parameter for this model.**

### Reduction of Vocabulary

To improve efficiency, we reduce the vocabulary to the top 200 most informative words per class, based on log-likelihood differences (spam_ll - ham_ll). Positive scores indicate spam-indicative words; negative for ham. This feature selection simplifies the model but may trade off some accuracy, as seen in the evaluation results.

In [16]:
scores = {}

model = models[0.5]
log_likelihood = model["log_likelihood"]

for word in V:
    spam_ll = log_likelihood["spam"][word]
    ham_ll = log_likelihood["ham"][word]

    scores[word] = spam_ll - ham_ll

top_spam_words = sorted(scores.items(), key=lambda x: x[1], reverse=True)[:200]
top_ham_words = sorted(scores.items(), key=lambda x: x[1])[:200]

print("Top spam words:")
for w, s in top_spam_words[:20]:
    print(w, s)

print("\nTop ham words:")
for w, s in top_ham_words[:20]:
    print(w, s)

Top spam words:
edua 8.141083426076886
devmail 7.687827654132373
telesp 7.622383745879989
foxmail 7.487156958670222
writely 7.427066382260338
sjk 7.2788599154730065
ephedra 7.275137826808632
discreet 7.2651439897262975
hoodia 7.252509582874527
revitalizes 6.9226271559323695
zipping 6.9226271559323695
bestsellers 6.860620058158197
audition 6.833926385515719
greylink 6.831992148935888
mediadescription 6.831992148935888
midsize 6.831992148935888
shopzilla 6.831992148935888
titlelink 6.831992148935888
aspirin 6.780342139381612
caffeine 6.77830132234676

Top ham words:
handyboard -8.00217545841971
dmdx -7.817538653569619
lugnet -7.052317528413277
motors -6.763670670551398
sensor -6.69715679771727
anybodyi -6.688869457468699
deagol -6.6607382072183405
lnx -6.616999457155432
gso -6.611020366099374
voltage -6.520048587893647
jcf -6.516759111243248
sensors -6.49679172572938
nodes -6.479842167415606
node -6.448589623911502
lcd -6.427198433930185
installed -6.4235883217060845
analog -6.3979458910

In [17]:
selected_words = set()

for w, _ in top_spam_words:
    selected_words.add(w)

for w, _ in top_ham_words:
    selected_words.add(w)

In [18]:
vocab_reduced = set()
word_counts_reduced = {
    "spam": Counter(),
    "ham": Counter()
}

for text, label in zip(X_train, y_train):
    words = set(word_level_tokenize(text))
    filtered_words = [w for w in words if w in selected_words]

    vocab_reduced.update(filtered_words)

    word_counts_reduced[label].update(filtered_words)

V_reduced = sorted(vocab_reduced)

In [19]:
best_lambda = 0.5

model_reduced = train_naive_bayes(
    word_counts_reduced,
    V_reduced,
    D_spam,
    D_ham,
    best_lambda
)

In [20]:
results = evaluate(
    X_test, y_test,
    model_reduced
)

print(results)
print("Precision:", results["precision"])
print("Recall:", results["recall"])
print("f1_score:", results["f1_score"])

{'TP': 7549, 'FP': 745, 'FN': 10, 'TN': 3043, 'precision': 0.910176030865686, 'recall': 0.9986770736869957, 'f1_score': 0.9523749448053996}
Precision: 0.910176030865686
Recall: 0.9986770736869957
f1_score: 0.9523749448053996


In [21]:
print("Vocabulary size before reduction:", len(V))
print("Vocabulary size after reduction:", len(V_reduced))

Vocabulary size before reduction: 125775
Vocabulary size after reduction: 400


After restricting the vocabulary to the most informative words, the recall slightly improved but the precision decreased. This occurs because the classifier loses many weaker but useful indicators of ham messages, causing more false positives.

Although the model becomes simpler, the overall F1-score decreases from the full-vocabulary classifier.

This shows that while informative words are powerful features, using the full vocabulary provides better classification performance.

In [ ]:
import pickle

with open("naive_bayes_model.pkl", "wb") as f:
    pickle.dump(model, f)

with open("naive_bayes_model_reduced.pkl", "wb") as f:
    pickle.dump(model_reduced, f)